README
Module: Text::CSV::Unicode

Description:
    Text::CSV::Unicode builds on Text::CSV to provides facilities for the 
    composition and decomposition of comma-separated values.  
    Text::CSV::Unicode handles input that includes 8-bit (and more) data.

    It is based on version 0.01 of Text::CSV, but later versions of that
    module are not compatible with Text::CSV::Unicode, so Text-CSV-0.01
    has been integrated into this distribution as lib/Text/CSV/Base.pm 
    and t/base.t; specifically
	% perl -pwe 's/Text::CSV/Text::CSV::Base/ && s/ {6}\#/\#/' \ 
		../Text-CSV-0.01/CSV.pm  > lib/Text/CSV/Base.pm
	% perl -pwe 's/Text::CSV/Text::CSV::Base/'		   \
		../Text-CSV-0.01/test.pl > t/base.t

Copying:
    Copyright (c) 2007, 2008, 2010, 2011, 2012 Robin Barker.  
    All rights reserved.

    This program is free software; you can redistribute it
    and/or modify it under the same terms as Perl itself.

    lib/Text/CSV/Base.pm (version 0.1) was a direct copy of 
    Text-CSV-0.01/CSV.pm, and t/base.t was a direct copy of
    Text-CSV-0.01/test.pl, with only the module name changed
    (and white-space changes), and these files are
    Copyright (c) 1997 Alan Citterman.

Author:
    Robin Barker <rmbarker@cpan.org>

    based on Text::CSV by
    Alan Citterman <alan@mfgrtl.com>

Prerequisites:
    perl 5.8.0

Build/Installation:
    Standard build/installation supported by ExtUtils::MakeMaker(3)...
	perl Makefile.PL
	make
	make test
	make install

Recent Changes:
    Version 0.114 2012/02/06
	strictness in tests

    Version 0.113 2011/08/24
	Text::CSV::Base 0.041
	allow undef input values in combine, without warnings

    Version 0.112 2010/02/17
	change to align with perl 5.12 unicode properties

    Version 0.111 2008/04/08
	changes to improve kwalite (see cpants.perl.org)

    Version 0.110 2008/01/01
	allow unicode in combine



Changes
Revision history for Perl extension Text::CSV::Unicode.

0.01  Tue Jul 10 14:21:57 2007
	- original version; created from my locally modified Text::CSV

0.02  2007-07-19
	- fix use of C< binary => 1 > in POD
	- added requires => { perl => 5.008 } to Build.PL

0.03  2007-07-24
	- Add C<use 5.008;> to Makefile.PL to prevent test failures

0.04  2007-08-08
	- two typos in lib/Text/CSV/Unicode.pm, and hide URL
	- skip_all tests in t/1.t if not perl v5.8.0 or Text::CSV

0.05  2007-09-25
	- use import to check Text::CSV 0.01 not the "UNAUTHORIZED RELEASE"
	- use a variable for \N{QUOTATION MARK} in strings and regexp

0.06  2007-12-30
	- remove inc/Module/Build/Compat.pm: functionality now in release
	- reinstate C<no warnings portable>: warnings clean for perl 5.10.0

****  Release of Text::CSV 1.00 (November 2007) broke 0.06 

0.099 2007-12-31
	- Text::CSV::Base to substitute for Text::CSV 0.01

0.100 2007-12-31
	- more documentation from Text::CSV 
	- documentation of the inclusion of lib/Text/CSV/Base.pm t/base.t

0.110 2008-01-01
	- use {_CHAROK} in combine
	- fix Text::CSV::Unicode->version
	- fix documentation of allowable characters in CAVEATS

0.111 2008-04-08
    	Changes to improve kwalitee
	- Build.PL not executable
	- added C<use warnings> to Text::CSV::Base
	- added Test::More to build_requires
	- added examples/

0.112 2010-02-17
	Change to _CHAROK due to changes to unicode properties in perl 5.11.4

0.113 2011-08-24
	Update Text/CSV/Base.pm to version 0.041
	Add t/combine-undef.t
	- avoid warnings on undef values in input array to combine

0.114 2012-02-06
	Some testers were failing with C<use strict> type failures
	in the test scripts, despite no C<use strict>
	- add <use strict;> and make the test scripts strict



lib/Text/CSV/Unicode.pm
package Text::CSV::Unicode;

# $Date: 2012-02-06 15:28:09 +0000 (Mon, 06 Feb 2012) $
# $Revision: 285 $
# $Source: $
# $URL: $

use 5.008;
use strict;
use warnings;
use Text::CSV::Base ();
use base qw(Text::CSV::Base);

our $VERSION = '0.114';

sub new {
    my $self = shift->SUPER::new();
    my %opts =
      ( @_ == 1 and ref $_[0] and ( ref $_[0] ) eq 'HASH' ) ? %{ $_[0] } : @_;
    $self->{'_CHAROK'} =
      $opts{binary} ? qr{ \p{Any} }msx : qr{ \t | \P{Cntrl} }msx;
    return $self;
}

1;

__END__

=head1 NAME

Text::CSV::Unicode -	comma-separated values manipulation routines
			with potentially wide character data

=head1 SYNOPSIS

    use Text::CSV::Unicode;

    $csv = Text::CSV::Unicode->new( { binary => 1 } );

    # then use methods from Text::CSV::Base (= Text::CSV 0.01)

    $version = Text::CSV::Unicode->version();	# get the module version

    $csv = Text::CSV::Unicode->new();	# create a new object

    $status = $csv->combine(@columns);	# combine columns into a string
    $line = $csv->string();		# get the combined string

    $status = $csv->parse($line);	# parse a CSV string into fields
    @columns = $csv->fields();		# get the parsed fields

    $status = $csv->status();		# get the most recent status
    $bad_argument = $csv->error_input();# get the most recent bad argument

=head1 DESCRIPTION

Text::CSV::Unicode provides facilities for the composition and
decomposition of comma-separated values, based on Text::CSV 0.01.
Text::CSV::Unicode allows for input with wide character data.

An instance of the Text::CSV::Unicode class can combine fields
into a CSV string and parse a CSV string into fields.

=head1 FUNCTIONS

=over 4

=item version

    $version = Text::CSV::Unicode->version();

This function may be called as a class or an object method. 
It returns the current module version.

=item new

    $csv = Text::CSV::Unicode->new( [{ binary => 1 }] );

This function may be called as a class or an object method.
It returns a reference to a newly created Text::CSV::Unicode object.
C<< binary => 0 >> allows the same ASCII input as Text::CSV and all
other input, while C<< binary => 1 >> allows for all Unicode
characters in the input (including \r and \n),

=item combine

    $status = $csv->combine(@columns);

This object function constructs a CSV string from the arguments,
returning success or failure.  Failure can result from lack of
arguments or an argument containing an invalid character.  Upon
success, C<string()> can be called to retrieve the resultant CSV
string.  Upon failure, the value returned by C<string()> is undefined
and C<error_input()> can be called to retrieve an invalid argument.

Silently accepts undef values in input and treats as an empty string.

=item string

    $line = $csv->string();

This object function returns the input to C<parse()> or the resultant
CSV string of C<combine()>, whichever was called more recently.

=item parse

    $status = $csv->parse($line);

This object function decomposes a CSV string into fields, returning
success or failure.  Failure can result from a lack of argument or
the given CSV string is improperly formatted.  Upon success,
C<fields()> can be called to retrieve the decomposed fields.  
Upon failure, the value returned by C<fields()> is undefined
and C<error_input()> can be called to retrieve the invalid argument.

=item fields

    @columns = $csv->fields();

This object function returns the input to C<combine()> or the resultant
decomposed fields of C<parse()>, whichever was called more recently.

=item status

    $status = $csv->status();

This object function returns success (or failure) of C<combine()> or
C<parse()>, whichever was called more recently.

=item error_input

    $bad_argument = $csv->error_input();

This object function returns the erroneous argument (if it exists) of
C<combine()> or C<parse()>, whichever was called more recently.

=back

=head1 SUBROUTINES/METHODS

None

=head1 DIAGNOSTICS

None

=head1 CONFIGURATION AND ENVIRONMENT

See HASH option to C<< ->new >>.

=head1 DEPENDENCIES

perl 5.8.0

=head1 INCOMPATIBILITIES

None

=head1 BUGS AND LIMITATIONS

As slow as Text::CSV 0.01.

Cannot change separators and delimiters.

=head1 EXAMPLE

    require Text::CSV::Unicode;

    my $csv = Text::CSV::Unicode->new;

    my $column = '';
    my $sample_input_string = '"I said, ""Hi!""",Yes,"",2.34,,"1.09"';
    if ($csv->parse($sample_input_string)) {
	my @field = $csv->fields;
	my $count = 0;
	for $column (@field) {
	    print ++$count, " => ", $column, "\n";
	}
	print "\n";
    }
    else {
	my $err = $csv->error_input;
	print "parse() failed on argument: ", $err, "\n";
    }

    my @sample_input_fields = ( 'You said, "Hello!"',
				5.67,
				'Surely',
				'',
				'3.14159');
    if ($csv->combine(@sample_input_fields)) {
	my $string = $csv->string;
	print $string, "\n";
    }
    else {
	my $err = $csv->error_input;
	print "combine() failed on argument: ", $err, "\n";
    }

=head1 CAVEATS

This module is based upon a working definition of CSV format
which may not be the most general.

=over 4

=item 1 

Allowable characters within a CSV field are all unicode characters,
with C<< binary => 1 >>; otherwise control characters are not allowed,
but the tab character is allowed.

=item 2

A field within CSV may be surrounded by double-quotes.

=item 3

A field within CSV must be surrounded by double-quotes to contain a comma.

=item 4

A field within CSV must be surrounded by double-quotes to contain an embedded
double-quote, represented by a pair of consecutive double-quotes.

=item 5

Line-ending characters are handled as part of the data.

=back

=head1 VERSION

0.111

=head1 AUTHOR

Robin Barker <rmbarker@cpan.org>

=head1 SEE ALSO

Text::CSV 0.01

=head1 LICENSE AND COPYRIGHT

Copyright (c) 2007, 2008, 2010, 2011, 2012 Robin Barker.  
All rights reserved. 

This program is free software; you can redistribute it
and/or modify it under the same terms as Perl itself.

The documentation of Text::CSV::Unicode methods that are inherited from
Text::CSV::Base is taken from Text::CSV 0.01 (with some reformatting) 
and is Copyright (c) 1997 Alan Citterman.

=cut



lib/Text/CSV/Base.pm
package Text::CSV::Base;

# Copyright (c) 1997 Alan Citterman. All rights reserved.
# This program is free software; you can redistribute it and/or
# modify it under the same terms as Perl itself.

################################################################################
# HISTORY
#
# Written by:
#    Alan Citterman <alan@mfgrtl.com>
#
# Version 0.01  06/05/1997
#    original version
#
# Maintained (as Text::CSV::Base) by:
#    Robin Barker <rmbarker@cpan.org>
#
# Version 0.01  2007-12-30
#    converted to Text::CSV::Base
#	% perl -pwe 's/Text::CSV/Text::CSV::Base/ && s/ {6}\#/\#/' \ 
#		../Text-CSV-0.01/CSV.pm  > lib/Text/CSV/Base.pm
#
# Version 0.02  2007-12-31
#    correct POD; update BEGIN {}
#
# Version 0.03  2008-01-01
#    make ->version an inheritable package method
#    remove AutoLoader
#    use _CHAROK in combine
################################################################################
# $Date: 2008-01-01 14:54:18 +0000 (Tue, 01 Jan 2008) $
# $Revision: 203 $
# $Source: $
# $URL: $

require 5.006;

use strict;

our $VERSION = '0.03';

# PBP does not like "\042"
use charnames qw(:full);
my $quote = "\N{QUOTATION MARK}";
my $qqr   = qr{ $quote }msx;

################################################################################
# version
#
#    class/object method expecting no arguments and returning the version number
#    of Text::CSV::Base.  there are no side-effects.
################################################################################
sub version {
    my $self = shift;
    my $vers = $self -> VERSION();
    return $vers if defined $vers;

    # inherit missing version from parent: 
    # pass empty subclass tests
    return eval {
	no strict qw(refs);
	((ref $self or $self).'::ISA') -> [0] -> version();
    }
}

################################################################################
# new
#
#    class/object method expecting no arguments and returning a reference to a
#    newly created Text::CSV::Base object.
################################################################################
sub new {
  my $proto = shift;
  my $class = ref($proto) || $proto;
  my $self = {};
  $self->{'_STATUS'} = undef;
  $self->{'_ERROR_INPUT'} = undef;
  $self->{'_STRING'} = undef;
  $self->{'_FIELDS'} = undef;

  $self->{'_CHAROK'} = qr{[\t\040-\176]};
# This is the regexp from the line now marked RMB (in _bite)

  bless $self, $class;
  return $self;
}

################################################################################
# status
#
#    object method returning the success or failure of the most recent combine()
#    or parse().  there are no side-effects.
################################################################################
sub status {
  my $self = shift;
  return $self->{'_STATUS'};
}

################################################################################
# error_input
#
#    object method returning the first invalid argument to the most recent
#    combine() or parse().  there are no side-effects.
################################################################################
sub error_input {
  my $self = shift;
  return $self->{'_ERROR_INPUT'};
}

################################################################################
# string
#
#    object method returning the result of the most recent combine() or the
#    input to the most recent parse(), whichever is more recent.  there are no
#    side-effects.
################################################################################
sub string {
  my $self = shift;
  return $self->{'_STRING'};
}

################################################################################
# fields
#
#    object method returning the result of the most recent parse() or the input
#    to the most recent combine(), whichever is more recent.  there are no
#    side-effects.
################################################################################
sub fields {
  my $self = shift;
  if (ref($self->{'_FIELDS'})) {
    return @{$self->{'_FIELDS'}};
  }
  return undef;
}

################################################################################
# combine
#
#    object method returning success or failure.  the given arguments are
#    combined into a single comma-separated value.  failure can be the result of
#    no arguments or an argument containing an invalid character.   side-effects
#    include:
#      setting status()
#      setting fields()
#      setting string()
#      setting error_input()
################################################################################
sub combine {
  my $self = shift;
  my @part = @_;
  $self->{'_FIELDS'} = \@part;
  $self->{'_ERROR_INPUT'} = undef;
  $self->{'_STATUS'} = 0;
  $self->{'_STRING'} = '';
  my $column = '';
  my $combination = '';
  my $skip_comma = 1;
  my $char_ok = qr{ \A $self->{_CHAROK} * \z }msx; 	# RMB
  if ($#part >= 0) {

    # at least one argument was given for "combining"...
    for $column (@part) {
      if ($column !~ $char_ok) {			# RMB

	# an argument contained an invalid character...
	$self->{'_ERROR_INPUT'} = $column;
	return $self->{'_STATUS'};
      }
      if ($skip_comma) {

	# do not put a comma before the first argument...
	$skip_comma = 0;
      } else {

	# do put a comma before all arguments except the first argument...
	$combination .= ',';
      }
      $column =~ s/\042/\042\042/go;
      $combination .= "\042";
      $combination .= $column;
      $combination .= "\042";
    }
    $self->{'_STRING'} = $combination;
    $self->{'_STATUS'} = 1;
  }
  return $self->{'_STATUS'};
}

################################################################################
# parse
#
#    object method returning success or failure.  the given argument is expected
#    to be a valid comma-separated value.  failure can be the result of
#    no arguments or an argument containing an invalid sequence of characters.
#    side-effects include:
#      setting status()
#      setting fields()
#      setting string()
#      setting error_input()
################################################################################
sub parse {
  my $self = shift;
  $self->{'_STRING'} = shift;
  $self->{'_FIELDS'} = undef;
  $self->{'_ERROR_INPUT'} = $self->{'_STRING'};
  $self->{'_STATUS'} = 0;
  if (!defined($self->{'_STRING'})) {
    return $self->{'_STATUS'};
  }
  our $parse_char_ok = qr{ \A $self->{_CHAROK} }msx; 	# RMB
  my $keep_biting = 1;
  my $palatable = 0;
  my $line = $self->{'_STRING'};
  if ($line =~ /\n$/) {
    chop($line);
    if ($line =~ /\r$/) {
      chop($line);
    }
  }
  my $mouthful = '';
  my @part = ();
  while ($keep_biting and ($palatable = $self->_bite(\$line, \$mouthful, \$keep_biting))) {
    push(@part, $mouthful);
  }
  if ($palatable) {
    $self->{'_ERROR_INPUT'} = undef;
    $self->{'_FIELDS'} = \@part;
  }
  return $self->{'_STATUS'} = $palatable;
}

################################################################################
# _bite
#
#    *private* class/object method returning success or failure.  the arguments
#    are:
#      - a reference to a comma-separated value string
#      - a reference to a return string
#      - a reference to a return boolean
#    upon success the first comma-separated value of the csv string is
#    transferred to the return string and the boolean is set to true if a comma
#    followed that value.  in other words, "bite" one value off of csv
#    returning the remaining string, the "piece" bitten, and if there's any
#    more.  failure can be the result of the csv string containing an invalid
#    sequence of characters.
#   
#    from the csv string and 
#    to be a valid comma-separated value.  failure can be the result of
#    no arguments or an argument containing an invalid sequence of characters.
#    side-effects include:
#      setting status()
#      setting fields()
#      setting string()
#      setting error_input()
################################################################################
# This subroutine is copied verbatim from Text::CSV Version 0.01
# and is Copyright (c) 1997 Alan Citterman.
# Robin Barker changed one line - marked RMB below
# - but then applied Perl Best Practices to code.

sub _bite {
    my ( $self, $line_ref, $piece_ref, $bite_again_ref ) = @_;
    my $in_quotes = 0;
    my $ok        = 0;
    ${$piece_ref}      = q{};
    ${$bite_again_ref} = 0;

    while (1) {
        if ( length( ${$line_ref} ) < 1 ) {

            # end of string...
            if ($in_quotes) {

                # end of string, missing closing double-quote...
            }
            else {

                # proper end of string...
                $ok = 1;
            }
            last;
        }
        if ( ${$line_ref} =~ m{\A $qqr }mosx ) {

            # double-quote...
            if ($in_quotes) {
                if ( length( ${$line_ref} ) == 1 ) {

                    # closing double-quote at end of string...
                    substr ${$line_ref}, 0, 1, q{};
                    $ok = 1;
                    last;
                }
                elsif ( ${$line_ref} =~ m{\A $qqr {2} }mosx ) {

                    # an embedded double-quote...
                    ${$piece_ref} .= $quote;
                    substr ${$line_ref}, 0, 2, q{};
                }
                elsif ( ${$line_ref} =~ m{\A $qqr , }mosx ) {

                    # closing double-quote followed by a comma...
                    substr ${$line_ref}, 0, 2, q{};
                    ${$bite_again_ref} = 1;
                    $ok = 1;
                    last;
                }
                else {

                    # double-quote, followed by undesirable character
                    # (bad character sequence)...
                    last;
                }
            }
            else {
                if ( length( ${$piece_ref} ) < 1 ) {

                    # starting double-quote at beginning of string
                    $in_quotes = 1;
                    substr ${$line_ref}, 0, 1, q{};
                }
                else {

                    # double-quote, outside of double-quotes
                    # (bad character sequence)...
                    last;
                }
            }
        }
        elsif ( ${$line_ref} =~ m{\A , }msx ) {

            # comma...
            if ($in_quotes) {

                # a comma, inside double-quotes...
                ${$piece_ref} .= substr ${$line_ref}, 0, 1, q{};
            }
            else {

                # a comma, which separates values...
                substr ${$line_ref}, 0, 1, q{};
                ${$bite_again_ref} = 1;
                $ok = 1;
                last;
            }
        }
        elsif ( ${$line_ref} =~ our $parse_char_ok ) {	# RMB: changed line

            # a tab, space, or printable...
            ${$piece_ref} .= substr ${$line_ref}, 0, 1, q{};
        }
        else {

            # an undesirable character...
            last;
        }
    }
    return $ok;
}

1;

__END__

=head1 NAME

Text::CSV::Base - comma-separated values manipulation routines

=head1 SYNOPSIS

 use Text::CSV::Base;

 $version = Text::CSV::Base->version();# get the module version

 $csv = Text::CSV::Base->new();        # create a new object

 $status = $csv->combine(@columns);    # combine columns into a string
 $line = $csv->string();               # get the combined string

 $status = $csv->parse($line);         # parse a CSV string into fields
 @columns = $csv->fields();            # get the parsed fields

 $status = $csv->status();             # get the most recent status
 $bad_argument = $csv->error_input();  # get the most recent bad argument

=head1 DESCRIPTION

Text::CSV::Base provides facilities for the composition and decomposition of
comma-separated values.  An instance of the Text::CSV::Base class can combine
fields into a CSV string and parse a CSV string into fields.

=head1 FUNCTIONS

=over 4

=item version

 $version = Text::CSV::Base->version();

This function may be called as a class or an object method.  It returns the current
module version.

=item new

 $csv = Text::CSV::Base->new();

This function may be called as a class or an object method.  It returns a reference to a
newly created Text::CSV::Base object.

=item combine

 $status = $csv->combine(@columns);

This object function constructs a CSV string from the arguments, returning
success or failure.  Failure can result from lack of arguments or an argument
containing an invalid character.  Upon success, C<string()> can be called to
retrieve the resultant CSV string.  Upon failure, the value returned by
C<string()> is undefined and C<error_input()> can be called to retrieve an
invalid argument.

=item string

 $line = $csv->string();

This object function returns the input to C<parse()> or the resultant CSV string of
C<combine()>, whichever was called more recently.

=item parse

 $status = $csv->parse($line);

This object function decomposes a CSV string into fields, returning
success or failure.  Failure can result from a lack of argument or the given CSV
string is improperly formatted.  Upon success, C<fields()> can be called to
retrieve the decomposed fields .  Upon failure, the value returned by
C<fields()> is undefined and C<error_input()> can be called to retrieve the
invalid argument.

=item fields

 @columns = $csv->fields();

This object function returns the input to C<combine()> or the resultant decomposed
fields of C<parse()>, whichever was called more recently.

=item status

 $status = $csv->status();

This object function returns success (or failure) of C<combine()> or C<parse()>,
whichever was called more recently.

=item error_input

 $bad_argument = $csv->error_input();

This object function returns the erroneous argument (if it exists) of C<combine()>
or C<parse()>, whichever was called more recently.

=back

=head1 EXAMPLE

  require Text::CSV::Base;

  my $csv = Text::CSV::Base->new;

  my $column = '';
  my $sample_input_string = '"I said, ""Hi!""",Yes,"",2.34,,"1.09"';
  if ($csv->parse($sample_input_string)) {
    my @field = $csv->fields;
    my $count = 0;
    for $column (@field) {
      print ++$count, " => ", $column, "\n";
    }
    print "\n";
  } else {
    my $err = $csv->error_input;
    print "parse() failed on argument: ", $err, "\n";
  }

  my @sample_input_fields = ('You said, "Hello!"',
			     5.67,
			     'Surely',
			     '',
			     '3.14159');
  if ($csv->combine(@sample_input_fields)) {
    my $string = $csv->string;
    print $string, "\n";
  } else {
    my $err = $csv->error_input;
    print "combine() failed on argument: ", $err, "\n";
  }

=head1 CAVEATS

This module is based upon a working definition of CSV format which may not be
the most general.

=over 4

=item 1 

Allowable characters within a CSV field include 0x09 (tab) and the inclusive
range of 0x20 (space) through 0x7E (tilde).

=item 2

A field within CSV may be surrounded by double-quotes.

=item 3

A field within CSV must be surrounded by double-quotes to contain a comma.

=item 4

A field within CSV must be surrounded by double-quotes to contain an embedded
double-quote, represented by a pair of consecutive double-quotes.

=item 5

A CSV string may be terminated by 0x0A (line feed) or by 0x0D,0x0A
(carriage return, line feed).

=back

=head1 AUTHOR

Alan Citterman F<E<lt>alan@mfgrtl.comE<gt>>

=head1 SEE ALSO

perl(1)

=cut
